### ISSUE TITLE
stdin,stdout,stderr=ssh.exec_command('dis ver') , faced issue

### ISSUE THREAD

**Issue by cyb1129** [paramiko/paramiko#1304]
i write py for login Huawei switch  , as fllowss:
=================
import paramiko
ssh=paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
paramiko.util.log_to_file('syslogin.log')
ssh.connect('192.168.43.100',username='admin',password='admin123')
stdin,stdout,stderr=ssh.exec_command('dis ver')    #there pause forever,long long time. there is issue 
print(stdout.read())
===============
then i get log , find until here : 
DEB [20180928-22:49:20.662] thr=1   paramiko.transport: userauth is OK
INF [20180928-22:49:20.797] thr=1   paramiko.transport: Authentication (password) successful!
DEB [20180928-22:49:20.797] thr=2   paramiko.transport: [chan 0] Max packet in: 32768 bytes
DEB [20180928-22:49:20.930] thr=1   paramiko.transport: [chan 0] Max packet out: 32768 bytes
DEB [20180928-22:49:20.930] thr=1   paramiko.transport: Secsh channel 0 opened.

the next line should be :  Secsh channel 0 request  ok ,   
but not print to log ,  i cannot find any reason ,  anyone could hep me ?

and i had do it successfully  ssh to centos  , so this is also  strange .

**Comment by radssh** [comment]
Many network appliance/switch implementations of SSH server are not entirely compatible with all the RFC standards, and more specifically the OpenSSH implementation that Paramiko targets as the reference implementation.

Having does some testing with Cisco and other managed switches, I believe they did not behave well with the `exec_command` call, and required `get_pty()` and `invoke_shell()` as necessary setup, and `Channel.send()` to submit the command line to execute the command. This is much more effort compared to `exec_command()`, but it may be the only way to run commands via the paramiko connection.

You might be able to test whether you just need the `get_pty()` call before `exec_command()` using the ssh command line client. Typically when you include a command to run, it will not allocate a pty, and use the same protocol message as paramiko's exec_command. If this fails to work:
```
ssh -vv admin@192.168.43.100 dis ver
```
this one might:
```
ssh -t -vv admin@192.168.43.100 dis ver
```

If the second one does work properly, then you can probably still use `exec_command()`, but will have to explicitly call `get_pty()`, which is what the `-t` flag does for the ssh command line. If neither one works, then you will need both a `get_pty()` as well as `invoke_shell()` and use `send()` and `recv()`

**Comment by cyb1129** [comment]
> Many network appliance/switch implementations of SSH server are not entirely compatible with all the RFC standards, and more specifically the OpenSSH implementation that Paramiko targets as the reference implementation.
> 
> Having does some testing with Cisco and other managed switches, I believe they did not behave well with the `exec_command` call, and required `get_pty()` and `invoke_shell()` as necessary setup, and `Channel.send()` to submit the command line to execute the command. This is much more effort compared to `exec_command()`, but it may be the only way to run commands via the paramiko connection.
> 
> You might be able to test whether you just need the `get_pty()` call before `exec_command()` using the ssh command line client. Typically when you include a command to run, it will not allocate a pty, and use the same protocol message as paramiko's exec_command. If this fails to work:
> 
> ```
> ssh -vv admin@192.168.43.100 dis ver
> ```
> this one might:
> 
> ```
> ssh -t -vv admin@192.168.43.100 dis ver
> ```
> If the second one does work properly, then you can probably still use `exec_command()`, but will have to explicitly call `get_pty()`, which is what the `-t` flag does for the ssh command line. If neither one works, then you will need both a `get_pty()` as well as `invoke_shell()` and use `send()` and `recv()`

hi,thanks , i had get result by invoke_shell() successfully , but still have  issue , when i  use chan.resv() , there need (nbytes) ,  sometimes i  donot know  how many  message will receive ,  so there has any parameter could set  then no need input such as 1000000 as number

**Comment by radssh** [comment]
Here, you need to probably make use of `recv_ready()` to see if there are input bytes waiting, and check to see if the result leaves you back at the shell prompt. Once you use `invoke_shell()`, you lose the convenience that `exec_command()` gives you, and you are programmatically driving the shell session itself.

**Comment by cyb1129** [comment]
> Here, you need to probably make use of `recv_ready()` to see if there are input bytes waiting, and check to see if the result leaves you back at the shell prompt. Once you use `invoke_shell()`, you lose the convenience that `exec_command()` gives you, and you are programmatically driving the shell session itself.

yes, i could use send()  and recv() , but recv() request get nbytes ， there 2 questions:
1. i  cannot sure this command could back how many bytes ,  i set nbytes=1000000 or more , if back more than 1000000 , the rest of them wono't print , so i want to  it could print all 
2.i want to find some value such as '---More---', when i find it , i need send '\n' , until '---More---' not print , and get all print to file. 
as u know , stdin,stdout,stderr=ssh.exec_command() ,there is a session , but invoke_shell is channel ,
